src/Catalyst.Core.Modules.Consensus.Tests/UnitTests/Deltas/DeltaHubTests.cs

```diff
@@ -0,0 +1,203 @@
+
+            _hub.BroadcastCandidate(myCandidate);
+            _broadcastManager.Received(1).BroadcastAsync(Arg.Is<ProtocolMessage>(
+                m => IsExpectedCandidateMessage(m, myCandidate, _peerId)));
+        }
+        
+        [Fact]
+        public void BroadcastFavouriteCandidateDelta_Should_Broadcast()
+        {
+            var favourite = new FavouriteDeltaBroadcast
+            {
+                Candidate = DeltaHelper.GetCandidateDelta(_hashProvider),
+                VoterId = _peerId
+            };
+
+            _hub.BroadcastFavouriteCandidateDelta(favourite);
+            _broadcastManager.Received(1).BroadcastAsync(Arg.Is<ProtocolMessage>(
+                c => IsExpectedCandidateMessage(c, favourite, _peerId)));
+        }
+
+        [Fact]
+        public async Task PublishDeltaToIpfsAsync_should_return_ipfs_address()
+        {
+            var delta = DeltaHelper.GetDelta(_hashProvider);
+            var cid = CidHelper.CreateCid(_hashProvider.ComputeUtf8MultiHash("lskdjaslkjfweoho"));
+            var cancellationToken = new CancellationToken();
+
+            _dfs.AddAsync(Arg.Any<Stream>(), Arg.Any<string>(), cancellationToken).Returns(cid);
+
+            var deltaCid = await _hub.PublishDeltaToDfsAndBroadcastAddressAsync(delta, cancellationToken);
+            deltaCid.Should().NotBeNull();
+            deltaCid.Should().Be(cid);
+        }
+
+        [Fact]
+        public async Task PublishDeltaToIpfsAsync_should_retry_then_return_ipfs_address()
+        {
+            var delta = DeltaHelper.GetDelta(_hashProvider);
+            var cid = CidHelper.CreateCid(_hashProvider.ComputeUtf8MultiHash("success"));
+
+            var dfsResults = new SubstituteResults<Cid>(() => throw new Exception("this one failed"))
+               .Then(() => throw new Exception("this one failed too"))
+               .Then(cid);
+
+            _dfs.AddAsync(Arg.Any<Stream>(), Arg.Any<string>())
+               .Returns(ci => dfsResults.Next());
+
+            var deltaCid = await _hub.PublishDeltaToDfsAndBroadcastAddressAsync(delta);
+            deltaCid.Should().NotBeNull();
+            deltaCid.Should().Be(cid);
+
+            await _dfs.ReceivedWithAnyArgs(3).AddAsync(Arg.Any<Stream>(), Arg.Any<string>());
+        }
+
+        [Fact]
+        public async Task PublishDeltaToIpfsAsync_should_retry_until_cancelled()
+        {
+            var delta = DeltaHelper.GetDelta(_hashProvider);
+            var dfsHash = "success";
+            var cancellationSource = new CancellationTokenSource();
+            var cancellationToken = cancellationSource.Token;
+
+            var dfsResults = new SubstituteResults<string>(() => throw new Exception("this one failed"))
+               .Then(() => throw new Exception("this one failed again"))
+               .Then(() =>
+                {
+                    cancellationSource.Cancel();
+                    throw new Exception("this one failed too");
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: ['System.Exception' should not be thrown by user code.](https://app.codacy.com/app/catalyst-network/Catalyst.Node/pullRequest?prid=4318201)